backend/src/validation/serviceTokenDataV3.ts

```diff
@@ -0,0 +1,119 @@
+import { Types } from "mongoose";
+import { IServiceTokenDataV3 } from "../models";
+import { Permission } from "../models/serviceTokenDataV3";
+import { z } from "zod";
+import { UnauthorizedRequestError } from "../utils/errors";
+import { isValidScopeV3 } from "../helpers";
+import { AuthData } from "../interfaces/middleware";
+import { checkIPAgainstBlocklist } from "../utils/ip";
+
+/**
+ * Validate that service token (client) can access workspace
+ * with id [workspaceId] and its environment [environment] with required permissions
+ * [requiredPermissions]
+ * @param {Object} obj
+ * @param {ServiceTokenData} obj.serviceTokenData - service token client
+ * @param {Types.ObjectId} obj.workspaceId - id of workspace to validate against
+ * @param {String} environment - (optional) environment in workspace to validate against
+ * @param {String[]} acceptedPermissions - accepted permissions as part of the endpoint
+ */
+ export const validateServiceTokenDataV3ClientForWorkspace = async ({
+  authData,
+  serviceTokenData,
+  workspaceId,
+  environment,
+  secretPath = "/",
+  requiredPermissions
+}: {
+  authData: AuthData;
+  serviceTokenData: IServiceTokenDataV3;
+  workspaceId: Types.ObjectId;
+  environment?: string;
+  secretPath?: string;
+  requiredPermissions: Permission[];
+}) => {
+  
+  // validate ST V3 IP address
+  checkIPAgainstBlocklist({
+    ipAddress: authData.ipAddress,
+    trustedIps: serviceTokenData.trustedIps
+  });
+  
+  if (!serviceTokenData.workspace.equals(workspaceId)) {
+    // case: invalid workspaceId passed
+    throw UnauthorizedRequestError({
+      message: "Failed service token authorization for the given workspace"
+    });
+  }
+  
+  if (environment) {
```

Comment:
under which case will `environment` be empty? I think isValidScopeV3 should always run unless i'm missing something

Reply:
In the future case where ST V3 may be provisioned project-level actions that are not bound to a specific environment / path - not applicable to this PR specifically.